### PR TITLE
fix: withdraw limits

### PR DESCRIPTION
**Pull request by Schlagonia**
## Description

simulate the full withdraw flow to make sure maxRedeem and maxWithdraw fit the erc-4626 standard and vaults are compatible with each other.

Fixes # (issue)

## Checklist

- [ ] I have run vyper and solidity linting
- [ ] I have run the tests on my machine
- [ ] I have followed commitlint guidelines
- [ ] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

File: contracts/VaultV3.vy [hunk 1]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
-@view
-@internal
-def _max_withdraw(owner: address) -> uint256:
-    return self._convert_to_assets(self.balance_of[owner], Rounding.ROUND_DOWN)
+    This will do a full simulation of the withdraw in order to determine
+    how much is currently liquid and if the `max_loss` would allow for the 
+    tx to not revert.
+
+    This will track any expected loss to check if the tx will revert, but
+    not account for it in the amount returned since it is unrealised and
```
Comment (fp-crypto) [inline contracts/VaultV3.vy:567]: nit but `unrealised` is british english.
```suggestion
    not account for it in the amount returned since it is unrealized and 
```

File: contracts/VaultV3.vy [hunk 2]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
-    return self._convert_to_assets(self.balance_of[owner], Rounding.ROUND_DOWN)
+    This will do a full simulation of the withdraw in order to determine
+    how much is currently liquid and if the `max_loss` would allow for the 
+    tx to not revert.
+
+    This will track any expected loss to check if the tx will revert, but
+    not account for it in the amount returned since it is unrealised and 
+    therefore will not be accounted for in the conversion rates.
+
+    i.e. If we have 100 debt and 10 of unrealised loss, the max we can get
```
Comment (fp-crypto) [inline contracts/VaultV3.vy:570]:
```suggestion
    i.e. If we have 100 debt and 10 of unrealized loss, the max we can get
```

File: contracts/VaultV3.vy (hunk 1 above)
Comment (Schlagonia) [reply contracts/VaultV3.vy:567]: Yes, but thats whats used everywhere including the function name.

So I fear we are stuck using it in the comments too.

Unless we want to change all of them

File: contracts/VaultV3.vy [hunk 3]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+        loss: uint256 = 0
+
+        # If no queue was passed use the default one.
+        _strategies: DynArray[address, MAX_QUEUE] = strategies
+        if len(_strategies) == 0:
+            _strategies = self.default_queue
+
+        for strategy in _strategies:
+            # Can't use an invalid strategy.
+            assert self.strategies[strategy].activation != 0, "inactive strategy"
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:593]: Should not be. Is removed from the queue when revoked and checked before a custom queue is set
https://github.com/Schlagonia/yearn-vaults-v3/blob/limits/contracts/VaultV3.vy#L1288

File: contracts/VaultV3.vy [hunk 4]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+            loss += unrealised_loss
+
+            if needed == 0:
+                break
+
+        # Update the max after going through the queue.
+        max_assets = have
+
+        # Check if there is a loss and a non-default value was set.
+        if loss > 0 and max_loss < MAX_BPS:
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:629]: Its the default value used in "redeem" calls if one is not given.

It shouldnt be used. But dont think there is another option we should enforce as a default on every vault

File: contracts/VaultV3.vy [hunk 5]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+            # If the loss is not within the allowed range.
+            if loss > max_assets * max_loss / MAX_BPS:
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:631]: Well the actual redeem is <= so fits with what's used https://github.com/Schlagonia/yearn-vaults-v3/blob/limits/contracts/VaultV3.vy#L898

I think that makes sense. 'max loss' seems to be inclusive (only revert if greater than). 

If i set max_loss to 1% i would expect the tx to succeed if the actual loss == 1%

File: contracts/VaultV3.vy [hunk 6]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+
+            # See if any limit is enforced by the strategy.
+            strategy_limit: uint256 = IStrategy(strategy).maxWithdraw(self)
+
+            # Adjust accordingly if there is a max withdraw limit.
+            if strategy_limit < max_withdraw - unrealised_loss:
+                # lower unrealised loss to the proportion to the limit.
+                unrealised_loss = unrealised_loss * strategy_limit / max_withdraw
+                # Still count the unrealised loss as withdrawable.
+                max_withdraw = strategy_limit + unrealised_loss
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:609]: So this was an issue i was unsure how to handle. But I believe we have to count it as "withdrawable" since it is unrealized in order to comply with the standard.

I wrote the reasoning in the comments. https://github.com/Schlagonia/yearn-vaults-v3/blob/limits/contracts/VaultV3.vy#L566-L572

Because the conversion rates the vault uses dont yet know about the loss the amounts need to not adjust for it.

The maxRedeem would also be screwed up if we didnt account for the unrealized loss as withdrawable.

Its not an ideal way to do it but im pretty sure it has to be done this way to comply to the standard.

For a withdraw to be able to take on unrealized losses a custom non-zero 'max_loss' parameter must be used, so i would expect any integrators who pass in a non zero value to know its then possible to receive less than requested.

File: contracts/VaultV3.vy [hunk 7]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+            max_withdraw: uint256 = min(needed, self.strategies[strategy].current_debt)
+
+            # Get any unrealised loss for the strategy.
+            unrealised_loss: uint256 = self._assess_share_of_unrealised_losses(strategy, max_withdraw)
+
+            # See if any limit is enforced by the strategy.
+            strategy_limit: uint256 = IStrategy(strategy).maxWithdraw(self)
+
+            # Adjust accordingly if there is a max withdraw limit.
+            if strategy_limit < max_withdraw - unrealised_loss:
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:605]: Since .maxWithdraw(self) will already have accounted for the losses we cannot pass that value into the _assess_share_of_unrealised_losses. Otherwise our unrealised_losses will be reduced twice.

EX:

strategy_debt = 100
current_value = 90
unrealised_loss = 10

So if a user is trying to withdraw the full 100 the unrealised_loss should be 10.

strategy.maxWithdraw(vault) = 90
_assess_share_of_unrealised_losses(strategy, 90) = .1 * 90 = 9

**Comment by fp-crypto** [comment]
> simulate the full withdraw flow to max sure

simulate the full withdraw flow to **make**

File: contracts/VaultV3.vy (hunk 4 above)
[the remark this answers is not in the document]
Comment (fp-crypto) [reply contracts/VaultV3.vy:629]: Why is 99.999999999% a valid value, but not 100%?

[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:629]: 100% is a valid value and is used by default for redeems. But if its 100% then we have no need to waste the gas to check the loss is within the range

File: contracts/VaultV3.vy [hunk 8]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+        max_assets = have
+
+        # Check if there is a loss and a non-default value was set.
+        if loss > 0 and max_loss < MAX_BPS:
+            # If the loss is not within the allowed range.
+            if loss > max_assets * max_loss / MAX_BPS:
+                # The tx will revert. Can only withdraw idle.
+                max_assets = current_idle
+
+    return max_assets
```
Comment (fp-crypto) [inline contracts/VaultV3.vy:635]: This is underestimates the max_assets available in cases where partially iterating through the withdraw queue would have had tolerable losses.

Comment (Schlagonia) [reply contracts/VaultV3.vy:635]: https://github.com/yearn/yearn-vaults-v3/pull/179/commits/abb73b675c496a293a1562e792f1748aa011bf18

Added to the other branch with the modules.

I Did not account for a scenario in which partial would revert but the full queue would not. Think that we shouldnt return a value that has amounts less than it that would revert. Could break things.

Perhaps that could be something to add to a preivewWithdraw/redeem

File: contracts/VaultV3.vy [hunk 9]
```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
...
+
+        # Update the max after going through the queue.
+        max_assets = have
+
+        # Check if there is a loss and a non-default value was set.
+        if loss > 0 and max_loss < MAX_BPS:
+            # If the loss is not within the allowed range.
+            if loss > max_assets * max_loss / MAX_BPS:
+                # The tx will revert. Can only withdraw idle.
+                max_assets = current_idle
```
[the remark this answers is not in the document]
Comment (Schlagonia) [reply contracts/VaultV3.vy:633]: added here https://github.com/yearn/yearn-vaults-v3/pull/179/commits/abb73b675c496a293a1562e792f1748aa011bf18

**Comment by Schlagonia** [comment]
changes merged with this PR https://github.com/yearn/yearn-vaults-v3/commit/038718193aa88bfc55e45d4e1e6a7b24177b60eb

**Comment by Schlagonia** [comment]
changes merged with this PR https://github.com/yearn/yearn-vaults-v3/commit/038718193aa88bfc55e45d4e1e6a7b24177b60eb
